test/system/src/test/java/io/pravega/test/system/MultiReaderWriterTest.java

```diff
@@ -0,0 +1,267 @@
+        readWriteTest();
+    }
+
+    private void readWriteTest() throws InterruptedException, ExecutionException {
+
+        String scope = "testMultiReaderWriterScope" + new Random().nextInt(Integer.MAX_VALUE);
+        String readerGroupName = "testMultiReaderWriterReaderGroup" + new Random().nextInt(Integer.MAX_VALUE);
+
+        //20  readers -> 20 stream segments ( to have max read parallelism)
+        ScalingPolicy scalingPolicy = ScalingPolicy.fixed(5);
+        StreamConfiguration config = StreamConfiguration.builder().scope(scope)
+                .streamName(STREAM_NAME).scalingPolicy(scalingPolicy).build();
+
+        URI controllerUri = getControllerURI();
+        Controller controller = getController(controllerUri);
+
+        //create a scope
+        Boolean createScopeStatus = controller.createScope(scope).get();
+        log.debug("Create scope status {}", createScopeStatus);
+
+        //create a stream
+        Boolean createStreamStatus = controller.createStream(config).get();
+        log.debug("Create stream status {}", createStreamStatus);
+         eventsReadFromPravega = new ConcurrentLinkedQueue<>();
+
+         //stopWriteFlag = new AtomicBoolean(false);
+         stopReadFlag = new AtomicBoolean(false);
+         eventData = new AtomicLong(); //data used by each of the writers.
+         eventReadCount = new AtomicLong(); // used by readers to maintain a count of events.
+
+        log.info("events read from pravega {}", eventsReadFromPravega);
+        log.info("data written {}", eventData.get());
+        log.info("data read {}", eventReadCount.get());
+        log.info(" stop read flag {}", stopReadFlag);
+        ClientFactory clientFactory = getClientFactory(scope);
+
+        //start writing events to the stream with 20 writers
+        log.info("creating {} writers", NUM_WRITERS);
+        List<CompletableFuture<Void>> writerList = new ArrayList<>();
+        for (int i = 0; i < NUM_WRITERS; i++) {
+            log.info("starting writer{}", i);
+            writerList.add(startNewWriter(eventData, clientFactory));
+        }
+
+        log.info("Creating Reader group : {}", readerGroupName);
+        @Cleanup
+        ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(scope, controllerUri);
+        readerGroupManager.createReaderGroup(readerGroupName, ReaderGroupConfig.builder().startingTime(0).build(),
+                Collections.singleton(STREAM_NAME));
+
+        //create 20 readers
+        log.info("creating {} readers", NUM_READERS);
+        List<CompletableFuture<Void>> readerList = new ArrayList<>();
+        for (int i = 0; i < NUM_READERS; i++) {
+            log.info("starting reader{}", i);
+            readerList.add(startReader("reader"+i, clientFactory, readerGroupName,
+                    eventsReadFromPravega, eventData, eventReadCount, stopReadFlag));
+        }
+        while (NUM_EVENTS != eventsReadFromPravega.size()) {
```

Comment:
We need to wait until all writers have finished execution and only then check the number of events read.